### PR TITLE
Feature extraction

### PR DESCRIPTION
**Pull request by kirsten-stallings**
Added Feature Extraction section of code.

### CODE REVIEW

File: machine_learning_tutorial.ipynb [hunk 1]
```diff
@@ -984,80 +971,1205 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Balance your data\n",
+    "(We'll come back to this)\n",
+    "\n",
+    "Balancing data is important because sometimes we have so few examples of one class that the model scores great if we just pretend it never happens. Consider the model predicting if it is Christmas yet:\n",
+    "![Christmas](Christmas.jpg)\n",
+    "It is so rarely Christmas that it can guess that it never is. There are two main ways to [handle unbalanced data](https://towardsdatascience.com/handling-imbalanced-datasets-in-deep-learning-f48407a0e758):\n",
```
Comment (jambran) [inline machine_learning_tutorial.ipynb]: I really like this answer for this question: https://stats.stackexchange.com/questions/227088/when-should-i-balance-classes-in-a-training-data-set

Talk about when you should care about this and when you shouldn't

File: machine_learning_tutorial.ipynb [hunk 2]
```diff
@@ -984,80 +971,1205 @@
...
-    "### One-hot Encodings of Categorical Data"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "## Feature Extraction"
+    "\n",
+    "Just like you have to cut hotdogs into little bit-size chunks for toddlers, you have to chop data up into a way that a computer can understand and _learn_ from. We do this through one-hot coding and normalization.\n"
```
[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: It's about breaking a lot of information up into a way that the computer can handle. I'll think about it but leaving it for now.

File: machine_learning_tutorial.ipynb [hunk 3]
```diff
@@ -984,80 +971,1205 @@
...
+    "One hot encoding is how we turn categorical data into zeros and ones\n",
+    "![one-hot-encoding](figures\\one_hot_encoding.png)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "### Deriving categorical features"
+    "It seems obvious that text categorical should be one-hot encoded. But when should discrete. Let's think about the column \"N_Season\".... Does the fact that an episode is in season 4 mean that it is more _something_ as an episode in season 2? Does the fact that an episode had 4.4 million views tell you that it is more _something_ than an episode with 2.2 million views? If there is a relationship with how the discrete value increases or decreases, it should be treated like a number. If there is not, it should be one-hot-encoded."
```
[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: added!

File: machine_learning_tutorial.ipynb [hunk 4]
```diff
@@ -984,80 +971,1205 @@
...
+    "for i, row in data.iterrows():\n",
+    "    try:\n",
+    "        if data.at[i,\"N_serie\"] == data.at[i-1, \"N_serie\"] : #previous speaker in same episode\n",
+    "            data.at[i, f\"previous_speaker\"] = data.at[i-1, \"Name\"]\n",
+    "    except KeyError: #first index\n",
+    "        pass\n",
+    "    try:\n",
+    "        if data.at[i,\"N_serie\"] == data.at[i+1, \"N_serie\"] : #next speaker in same episode\n",
+    "            data.at[i, f\"next_speaker\"] = data.at[i+1, \"Name\"]\n",
+    "    except KeyError: #first index\n",
```
[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: yep! TY!

File: machine_learning_tutorial.ipynb [hunk 5]
```diff
@@ -984,80 +971,1205 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#now we'll create features for each of these words\n",
+    "for word in target_words:\n",
+    "    data[f\"says_{word}\"]=0 #initalize an empty column\n",
+    "for i, row in data.iterrows():\n",
```
[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: yeah. The issue is "apply" is a little unpredictable. I'll be sure to mention it though

File: machine_learning_tutorial.ipynb [hunk 6]
```diff
@@ -984,80 +971,1205 @@
...
+   "metadata": {},
+   "source": [
+    "## Normalizing your data\n",
+    "\n",
+    "(We'll come back to this)\n",
+    "\n",
+    "\n",
+    "Normalizing your data is when you transform numerical features so that they are between 0 and 1, inclusively. \n",
+    "\n",
+    "We do this because many machine learning algorithms rely on multiplying the features with a matrix of weights to get a value. If the number of views range from 5,000,000, to 20,000,000 while the rating range from 7.0 to 9.0, that large number of viewing will have a bigger impact on the model, even though we really just want the ratio of one episodes views to another."
```
[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: I'm adding this in but then skipping it until the "Refine" stage, given we have time. Just something to be familiar with.

[the remark this answers is not in the document]
Comment (kirsten-stallings) [reply machine_learning_tutorial.ipynb]: added this comment into the section though!

**Pull request closed without merging**